dist/src/main/java/io/camunda/webapps/controllers/ForwardErrorController.java

```diff
@@ -0,0 +1,125 @@
+import org.springframework.core.env.Environment;
+import org.springframework.security.authentication.AnonymousAuthenticationToken;
+import org.springframework.security.core.Authentication;
+import org.springframework.security.core.context.SecurityContextHolder;
+import org.springframework.security.web.util.matcher.AntPathRequestMatcher;
+import org.springframework.security.web.util.matcher.RequestMatcher;
+import org.springframework.stereotype.Controller;
+import org.springframework.web.bind.annotation.RequestMapping;
+
+@Controller
+public class ForwardErrorController {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(ForwardErrorController.class);
+
+  private static final String LOGIN_RESOURCE = "/api/login";
+  private static final String API = "/api/**";
+  private static final String PUBLIC_API = "/v*/**";
+  private static final String DEFAULT_MANAGEMENT_URI = "/actuator/**";
+  private static final String SSO_CALLBACK_URI = "/sso-callback";
+  private static final String NO_PERMISSION = "/noPermission";
+  private static final String IDENTITY_CALLBACK_URI = "/identity-callback";
+  private static final String REQUESTED_URL = "requestedUrl";
+
+  @Autowired private Environment environment;
+
+  private List<RequestMatcher> ignoredRequestsMatchers;
+
+  @PostConstruct
+  public void init() {
+    // list of endpoints for which 404 is returned
+    ignoredRequestsMatchers =
+        List.of(
+            new AntPathRequestMatcher(PUBLIC_API),
+            new AntPathRequestMatcher(API),
+            new AntPathRequestMatcher(NO_PERMISSION),
+            new AntPathRequestMatcher(SSO_CALLBACK_URI),
+            new AntPathRequestMatcher(IDENTITY_CALLBACK_URI),
+            // actuator endpoints
+            new AntPathRequestMatcher(
+                Optional.ofNullable(environment.getProperty("management.endpoints.web.base-path"))
+                    .map(s -> s.concat("/**"))
+                    .orElse(DEFAULT_MANAGEMENT_URI)));
+  }
+
+  @RequestMapping(value = {"/{regex:[\\w-]+}", "/**/{regex:[\\w-]+}"})
+  public String forward404(final HttpServletRequest request, final HttpServletResponse response) {
+    if (ignoredRequestsMatchers.stream()
+        .anyMatch(requestMatcher -> requestMatcher.matches(request))) {
+      response.setStatus(HttpServletResponse.SC_NOT_FOUND);
+      return null;
+    }
+    final String requestedURL = getRequestedURL(request);
+    if (!requestedURL.startsWith("/tasklist") && !requestedURL.startsWith("/operate")) {
+      if (isStandaloneTasklist()) {
+        return "redirect:/tasklist" + requestedURL;
+      } else {
+        return "redirect:/operate" + requestedURL;
+      }
+    } else if (isLoginDelegated() && isNotLoggedIn()) {
+      return saveRequestAndRedirectToLogin(request, requestedURL);
+    } else {
+      if (isStandaloneTasklist()) {
+        return "forward:/tasklist";
+      } else {
+        return "forward:/operate";
+      }
```

Comment:
In this case, the code can assume that the request URL starts with `tasklist` or `operate`, so it could use it and redirect to what is set in the URL (i.e., with what the URL starts). That way, the additional check if it is a standalone tasklist or not is not necessary.

Reply:
yes it makes sense 
besides that the forwarding in the last else block is not correct, we should forward to the requested app and the default one :)

Reply:
fixed!